provider/iterators.go

```diff
@@ -438,20 +438,54 @@ type csvIterator struct {
 	columnNames   []string
 	idx           int64
 	limit         int64
+	files         []filestore.Filepath
+	fileIdx       int64
+	store         FileStore
 }
 
 func (c *csvIterator) Next() bool {
 	if c.idx >= c.limit {
 		return false
 	}
 	row, err := c.reader.Read()
-	if err != nil {
-		if err == io.EOF {
+	// If we've reached the end of one file, we need to move to the next one
+	if err == io.EOF {
+		if c.fileIdx >= int64(len(c.files)) {
 			return false
-		} else {
+		}
+		b, err := c.store.Read(c.files[c.fileIdx])
+		if err != nil {
 			c.err = err
 			return false
 		}
+		c.fileIdx += 1
+
+		reader := csv.NewReader(bytes.NewReader(b))
+		headers, err := reader.Read()
+		if err != nil {
+			c.err = fferr.NewInternalError(err)
+			return false
+		}
+		// The column names should match across all files
+		if len(headers) != len(c.columnNames) {
+			c.err = fferr.NewInvalidArgumentError(fmt.Errorf("column names do not match across files"))
```

Comment:
I feel like this is the wrong error type, it's not an invalid argument it's a dataset error of some kind